internal/controller/apiextensions/composite/composed.go

```diff
@@ -133,6 +135,48 @@ func (r *APIDryRunRenderer) Render(ctx context.Context, cp resource.Composite, c
 	return errors.Wrap(r.client.Create(ctx, cd, client.DryRunAll), errName)
 }
 
+// A APIClaimConfigurator may use the API server to get or patch
+// the referenced composite claim
+type APIClaimConfigurator struct {
+	client client.Client
+}
+
+// NewAPIClaimConfigurator returns a APIClaimConfigurator that uses
+// the API server to get or patch the referenced composite claim.
+func NewAPIClaimConfigurator(c client.Client) *APIClaimConfigurator {
+	return &APIClaimConfigurator{client: c}
+}
+
+// GetClaim returns the composite claim referenced from the composite or nil if no
+// claim is referenced.
+func (r *APIClaimConfigurator) GetClaim(ctx context.Context, cp resource.Composite) (resource.CompositeClaim, error) {
+	claimRef := cp.GetClaimReference()
```

Comment:
One oddity my testing found was that this is not currently guaranteed to be set for all composites which have a claim. Unsure of whether this is a bug in the claim controller or the assumptions here, but the initial creation of a composite from the claim controller does not have the claim ref set and is instead updated later in the claim reconcile.

This leads to an extra reconcile by the claim and composite on initial claim creation.